services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/hakemus/Hakemus.kt

```diff
@@ -64,12 +96,43 @@ sealed interface HakemusData {
     fun toResponse(): HakemusDataResponse
 
     fun yhteystiedot(): List<Hakemusyhteystieto>
+
+    fun listChanges(other: HakemusData): List<String> {
+        if (this::class != other::class) {
+            throw IncompatibleHakemusDataException(this::class, other::class)
+        }
+
+        val changes = mutableListOf<String>()
+        checkChange(HakemusData::applicationType, other)?.let { changes.add(it) }
+        checkChange(HakemusData::name, other)?.let { changes.add(it) }
+        checkChange(HakemusData::startTime, other)?.let { changes.add(it) }
+        checkChange(HakemusData::endTime, other)?.let { changes.add(it) }
+        checkChange(HakemusData::paperDecisionReceiver, other)?.let { changes.add(it) }
+        checkChangeInYhteystieto(HakemusData::customerWithContacts, other)?.let { changes.add(it) }
+
+        val areas = areas
+        val otherAreas = other.areas
+        areas?.withIndex()?.forEach { (i, area) ->
+            checkChange(area, otherAreas?.getOrNull(i)) { "areas[$i]" }?.let { changes.add(it) }
+        }
+        if (otherAreas != null) {
+            if (areas == null) {
+                changes.addAll(otherAreas.withIndex().map { (i, _) -> "areas[$i]" })
+            } else {
+                otherAreas.withIndex().drop(areas.size).forEach { (i, area) ->
+                    checkChange(area, areas.getOrNull(i)) { "areas[$i]" }?.let { changes.add(it) }
+                }
+            }
+        }
+
```

Comment:
This logic might be better to be in its own function perhaps?

Also, the logic is a bit hard to follow for me, but it seems that if areas contains areas A, B and C and otherAreas contains areas D, B, C and F, the result would contains two times "areas[0]"; 1st comes from area A and 2nd from area F after dropping first 3 elements in otherAreas. Is this so and does it matter if it is so?

Reply:
No, I was wrong. `withIndex` makes sure that the index of the second change is 3.

Reply:
I think `areas.getOrNull(i)` on line 123 is always null since we dropped all common indexes. This means the `changeChange` always returns the given name string. It should be possible to simplify this greatly.

Reply:
Moving this to a separate method made handling the null lists a lot easier since they can be replaced by empty lists when calling.

I made some other simplifications after realizing that the call to `chckChange` is unnecessary, like directly filtering out the unchanged elements instead of mapping to `String?` and then filtering out the nulls.

The `checkChange` with a function parameter was only called from a single anymore, so I inlined the method.

I think this should be simpler to read now.